File: src/main/java/com/company/erp/core/response/PageResponse.java
package com.company.erp.core.response;

import org.springframework.data.domain.Page;

import java.util.List;
import java.util.function.Function;

/**
 * Page wrapper hiding Spring's 0-indexed internals. Clients see 1-indexed
 * page numbers per the README pagination convention.
 */
public record PageResponse<T>(
        List<T> items,
        int page,
        int pageSize,
        long totalItems,
        int totalPages
) {
    public static <T> PageResponse<T> from(Page<T> page) {
        return new PageResponse<>(
                page.getContent(),
                page.getNumber() + 1,
                page.getSize(),
                page.getTotalElements(),
                page.getTotalPages());
    }

    public static <S, T> PageResponse<T> from(Page<S> page, Function<S, T> transform) {
        return new PageResponse<>(
                page.getContent().stream().map(transform).toList(),
                page.getNumber() + 1,
                page.getSize(),
                page.getTotalElements(),
                page.getTotalPages());
    }
}

File: app/Support/Pagination/PageResponse.php
<?php

namespace App\Support\Pagination;

use Illuminate\Contracts\Pagination\LengthAwarePaginator;

/**
 * Page wrapper hiding the paginator internals; clients see 1-indexed page
 * numbers per the shared pagination convention. Port of the Spring
 * PageResponse record.
 */
final class PageResponse
{
    /**
     * @param LengthAwarePaginator $paginator
     * @param callable|null $transform  maps each item to its DTO/array form
     * @return array<string, mixed>
     */
    public static function from(LengthAwarePaginator $paginator, ?callable $transform = null): array
    {
        $items = $paginator->items();
        if ($transform !== null) {
            $items = array_map($transform, $items);
        }

        return [
            'items' => array_values($items),
            'page' => $paginator->currentPage(),
            'pageSize' => $paginator->perPage(),
            'totalItems' => $paginator->total(),
            'totalPages' => $paginator->lastPage(),
        ];
    }
}
